### PR TITLE
[mlir][transform] Handle multiple library preloading passes

### LINKED ISSUE #69111: Consider refactoring "Merge parsed libraries into one module"
On #68661 @matthias-springer  notes:
```
I think getPreloadedTransformModule should be doing that. It currently only takes the first preloaded library and ignores all others. Alternatively, this could also be done PreloadLibraryPass: if there is no module preloaded yet, add the first one. On every subsequent run of the PreloadLibraryPass pass (and for every file in the list of libraries), merge the symbols.
```

### PR DESCRIPTION
**Pull request by ingomueller-net**
The transform dialect stores a "library module" that the preload pass
can populate. Until now, each pass registered an additional module by
simply pushing it to a vector; however, the interpreter only used the
first of them. This commit turns the registration into "loading", i.e.,
each newly added module gets merged into the existing one. This allows
the loading to be split into several passes, and using the library in
the interpreter now takes all of them into account. While this design
avoids repeated merging every time the library is accessed, it requires
that the implementation of merging modules lives in the
`TransformDialect` target (since it at the dialect depend on each
other).

This resolves #69111.

### CODE REVIEW

File: mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h [hunk 1]
```diff
@@ -315,6 +315,62 @@ class BuildOnly : public DerivedTy {
   BuildOnly() : DerivedTy(/*buildOnly=*/true) {}
 };
 
+namespace detail {
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h:318]: Can this go to some new `Utils` header? This doesn't look like stuff that belongs to the dialect definition.

File: mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h [hunk 2]
```diff
@@ -315,6 +315,62 @@ class BuildOnly : public DerivedTy {
...
+
+/// Expands the given list of `paths` to a list of `.mlir` files.
+///
+/// Each entry in `paths` may either be a regular file, in which case it ends up
+/// in the result list, or a directory, in which case all (regular) `.mlir`
+/// files in that directory are added. Any other file types lead to a failure.
+LogicalResult expandPathsToMLIRFiles(ArrayRef<std::string> &paths,
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h:325]: Why take a reference to ArrayRef, which itself is a reference as the name indicates?

(I see the code is moved from elsewhere, but IIRC it's yours anyway).

File: mlir/lib/Dialect/Transform/IR/TransformDialect.cpp [hunk 3]
```diff
@@ -172,3 +195,357 @@ LogicalResult transform::TransformDialect::verifyOperationAttribute(
...
+  return success();
+}
+
+LogicalResult transform::detail::parseTransformModuleFromFile(
+    MLIRContext *context, llvm::StringRef transformFileName,
+    OwningOpRef<ModuleOp> &transformModule) {
+  if (transformFileName.empty()) {
+    LLVM_DEBUG(
+        DBGS() << "no transform file name specified, assuming the transform "
+                  "module is embedded in the IR next to the top-level\n");
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/IR/TransformDialect.cpp:258]: This message no longer corresponds to the usages of this function and is only confusing. Furthermore, I'm not convinced that returning success here is the right behavior for this function in the more general case. Nothing was parsed.

File: mlir/include/mlir/Dialect/Transform/IR/TransformDialect.td [hunk 4]
```diff
@@ -78,20 +78,19 @@ def Transform_Dialect : Dialect {
...
-    }
-
-    /// Returns a range of registered library modules.
-    auto getLibraryModules() const {
-      return ::llvm::map_range(
-          libraryModules,
-          [](const ::mlir::OwningOpRef<::mlir::ModuleOp> &library) {
-        return library.get();
-      });
+    /// Loads the given module into the transform symbol library module.
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/Transform/IR/TransformDialect.td:81]: Documentation looks copy-pasted from below.

File: mlir/lib/Dialect/Transform/IR/TransformDialect.cpp [hunk 5]
```diff
@@ -172,3 +195,357 @@ LogicalResult transform::TransformDialect::verifyOperationAttribute(
...
+  }
+
+  // Merge parsed libraries into one module.
+  auto loc = FileLineColLoc::get(context, "<shared-library-module>", 0, 0);
+  OwningOpRef<ModuleOp> mergedParsedLibraries =
+      ModuleOp::create(loc, "__transform");
+  {
+    mergedParsedLibraries.get()->setAttr("transform.with_named_sequence",
+                                         UnitAttr::get(context));
+    IRRewriter rewriter(context);
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/IR/TransformDialect.cpp:308]: Why do we have this?

File: mlir/lib/Dialect/Transform/IR/TransformDialect.cpp [hunk 6]
```diff
@@ -172,3 +195,357 @@ LogicalResult transform::TransformDialect::verifyOperationAttribute(
...
+
+  LLVM_DEBUG(DBGS() << "done merging ops\n");
+  return success();
+}
+
+LogicalResult transform::detail::mergeSymbolsInto(
+    Operation *target, MutableArrayRef<OwningOpRef<Operation *>> others) {
+  for (OwningOpRef<Operation *> &other : others) {
+    if (failed(transform::detail::mergeSymbolsInto(target, std::move(other))))
+      return target->emitError() << "failed to verify merged transform module";
```
Comment (ftynse) [inline mlir/lib/Dialect/Transform/IR/TransformDialect.cpp:548]: The callee has already reported an error, we shouldn't need to emit another error. It is also not necessarily a failure to verify.

**Review by ftynse — CHANGES_REQUESTED** [review]

File: mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h (hunk 1 above)
Comment (ingomueller-net) [reply mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h:318]: Yep, makes sense. I realized that I only have to move the `mergeSymbolsInto` function (and related static functions). The remaining ones (for expanding paths etc) are arguably more related to the interpreter pass than this one, so I am leaving them where they are.

For now, I am calling the files `Utils.h`/`Utils.cpp`. All other files are called `TransformSomething.xxx` -- should I rename the new files to `TransformUtils.xxx` as well to follow that scheme?

File: mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h (hunk 2 above)
Comment (ingomueller-net) [reply mlir/include/mlir/Dialect/Transform/IR/TransformDialect.h:325]: No good reason. I probably changed the type from something else and overlooked the `&` when I changed it to `ArrayRef`. Fixed in new PR.

File: mlir/lib/Dialect/Transform/IR/TransformDialect.cpp (hunk 3 above)
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/IR/TransformDialect.cpp:258]: True, good catch. I removed that whole check in the new PR.

File: mlir/lib/Dialect/Transform/IR/TransformDialect.cpp (hunk 5 above)
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/IR/TransformDialect.cpp:308]: Haha, no idea. Removed.

File: mlir/lib/Dialect/Transform/IR/TransformDialect.cpp (hunk 6 above)
Comment (ingomueller-net) [reply mlir/lib/Dialect/Transform/IR/TransformDialect.cpp:548]: This is being discussed in #69112/#69331. Can we follow up this discussion there?

**Review by ingomueller-net — COMMENTED** [review]
Thanks a lot for the review! I have addressed all raised issues here: #69705. I started a new PR because the code that I had to move has changed in `main` in the meantime, and I had to move it again (plus I decided to move fewer things). I'll abandon this PR in favor of the new one now.

**Comment by ingomueller-net** [comment]
Closing in favor of new attempt in #69705.

**Pull request closed without merging**